syntax = "proto3";

package TW.Substrate.Proto;
option java_package = "wallet.core.jni.proto";

message Era {
  // recent block number (called phase in polkadot code), should match block hash
  uint64 block_number = 1;

  // length of period, calculated from block number, e.g. 64
  uint64 period = 2;
}

message Balance {
  message Transfer {
    int32 module_index = 1;
    int32 method_index = 2;
    string to_address = 3;
    bytes value = 4; // big integer
  }
  message BatchTransfer {
    int32 module_index = 1;
    int32 method_index = 2;
    repeated Transfer transfers = 3;
  }
  oneof message_oneof {
    Transfer transfer = 1;
    BatchTransfer batchTransfer = 2;
  }
}

// Input data necessary to create a signed transaction.
message SigningInput {
  // Recent block hash, or genesis hash if era is not set
  bytes block_hash = 1;

  bytes genesis_hash = 2;

  // Current account nonce
  uint64 nonce = 3;

  uint32 spec_version = 4;
  uint32 transaction_version = 5;
  bytes tip = 6;  // big integer

  // Optional time validity limit, recommended, for replay-protection.  Empty means Immortal.
  Era era = 7;

  bytes private_key = 8;
  int32 network = 9;

  oneof message_oneof {
    Balance balance_call = 10;
  }
}

// Transaction signing output.
message SigningOutput {
  // Signed and encoded transaction bytes.
  bytes encoded = 1;
}

